---
title: "Assignment9"
author: "Alper Turgut"
date: "November 7, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rafalib)
library(Biobase)
library(Hmisc)
library(ggplot2)
library(UsingR)
library(contrast)
library(rafalib)
library(devtools)
library(sp)
library(limma)
library(tissuesGeneExpression)
library(GSE5859Subset)
library(genefilter)
library(qvalue)
library(RColorBrewer)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("hgfocus.db", version = "3.8")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("genefilter", version = "3.8")
```

# Batch Effects

## Exercises 1

### Load the admissions data from the dagdata package (which is available from the genomicsclass repository):
```{r}

install_github("genomicsclass/dagdata")
library(dagdata)
data(admissions)
```
### Familiarize yourself with this table:
```{r}
admissions
```
### 1. Let's compute the proportion of men who were accepted:
```{r}
index = which(admissions$Gender==1)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```
What is the proportion of women that were accepted?
```{r}
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

Proportion of women accepted=.3033

### 2. Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value?
```{r}
##make a 2 x 2 table
index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
menYes = sum(men$Number*men$Percent/100)
menNo = sum(men$Number*(1-men$Percent/100))
womenYes = sum(women$Number*women$Percent/100)
womenNo = sum(women$Number*(1-women$Percent/100))
tab = matrix(c(menYes,womenYes,menNo,womenNo),2,2)
print(chisq.test(tab)$p.val)
```
### This difference actually led to a lawsuit???. Now notice that looking at the data by major, the differences disappear. How can this be? This is referred to as Simpson's Paradox. In the following questions we will try to decipher why this is happening.

### 3. We can quantify how "hard" a major is by using the percent of students that were accepted.Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?
```{r}
H=cbind(admissions[1:6,c(1,3)],admissions[7:12,3])
colnames(H)[2:3]=c("Male","Female")
H
```

The hardest major is F.

### 4. What proportion is accepted for this major?
```{r}
index = which(admissions$Major=="F")
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied

```
Proportion = .06477591

### 5. For men, what is the correlation between the number of applications across majors and H?
```{r}
index<-which(admissions$Gender=="1")
numApps<-admissions$Number[index]
print(cor(numApps,H$Male))
```

The correlation is .7160985

### 6. For women, what is the correlation between the number of applications across majors and H?
```{r}
index<-which(admissions$Gender=="0")
numApps<-admissions$Number[index]
print(cor(numApps,H$Female))
```

The correlation is -.716.

### 7. Given the answers to the above, which best explains the differences in admission percentages when we combine majors? . A) We made a coding mistake when computing the overall admissions percentages. . B) There were more total number of women applications which made the denominator much bigger. . C) There is confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors. . D) The sample

Option C

## Exercises 2

### We will use the Bioconductor package Biobase which you can install with install_bioc function from rafalib: Load the data for this gene expression dataset:
```{r}
install_github("genomicsclass/GSE5859")
library(GSE5859)
data(GSE5859)
geneExpression = exprs(e)
sampleInfo = pData(e)
```

### 1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here. You can extract the year from each date like this:
```{r}
year = format(sampleInfo$date,"%y")
```
### Note that ethnic group and year is almost perfectly confounded:
```{r}
table(year,sampleInfo$ethnicity)
```
### 2. For how many of these years do we have more than one ethnicity represented?

2 years

### 3. Repeat the above exercise, but now, instead of year, consider the month as well. Specifically, instead of the year variable defined above use:
```{r}
month.year = format(sampleInfo$date,"%m%y")
```
### For what proportion of these month.year values do we have more than one ethnicity represented?
```{r}
month.year = format(sampleInfo$date,"%m%y")
myTable<-table(month.year,sampleInfo$ethnicity)
nrow(myTable)
myTable
```

1/21

### 4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene.
```{r}
cors <- cor(exprs(e))
Pairs=which(abs(cors)>0.9999,arr.ind=TRUE)
out = Pairs[which(Pairs[,1]<Pairs[,2]),,drop=FALSE]
if(length(out[,2])>0) e=e[,-out[2]]

out <- grep("AFFX",featureNames(e))
e <- e[-out,]

y <- exprs(e)-rowMeans(exprs(e))
dates <- pData(e)$date
eth <- pData(e)$ethnicity
annotation(e)
library(hgfocus.db)
annot <- select(hgfocus.db, keys=featureNames(e), keytype="PROBEID",
columns=c("CHR"))
##for genes with multiples, pick one
annot <-annot[match(featureNames(e),annot$PROBEID),]
annot$CHR <- ifelse(is.na(annot$CHR),NA,paste0("chr",annot$CHR))
##compute median expression on chromosome Y


index=which(sampleInfo$ethnicity%in%"CEU")
year = factor(format(dates,"%y"))[index]
year=droplevels(year,c("04","05","06"))
pvals<-rowttests(geneExpression[,index],year)$p.value
qvals<-qvalue(pvals)
print(sum(qvals$qvalues<.05))
```
Number of q-values is less than 4308

### 5. What is the estimate of pi0 provided by qvalue:
```{r}
print(qvals$pi0)
```

### 6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values less than 0.05?
```{r}
index<-which(sampleInfo$ethnicity=="CEU")
year <- factor(format(dates,"%y"))[index]
year<-droplevels(year,c("02","05","06"))
pvals<-rowttests(geneExpression[,index],year)$p.value
qvals<-qvalue(pvals)
print(sum(qvals$qvalues<.05))
```

### 7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05 ?
```{r}
index<-which(sampleInfo$ethnicity==c("CEU","ASN"))
ethnicity<-factor(sampleInfo$ethnicity)[index]
ethnicity<-droplevels(ethnicity,"HAN")
pvals<-rowttests(geneExpression[,index],ethnicity)$p.value
qvals<-qvalue(pvals)
print(sum(qvals$qvalues<.05))
```

### 8. Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05 ?
```{r}
dates<-format(dates,"%y")
index=which(dates=="05")
ethnicity=factor(sampleInfo$ethnicity)[index]
ethnicity=droplevels(ethnicity,"HAN")
pvals<-rowttests(geneExpression[,index],ethnicity)$p.value
qvals<-qvalue(pvals)
print(sum(qvals$qvalues<.05))
dates <- pData(e)$date
```
### 9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3). How many genes have q-values < 0.05 ?
```{r}
dates <- pData(e)$date
dates<-factor(format(dates,"%y"))
ethnicity=factor(sampleInfo$ethnicity)
set.seed(3)
index1 = which(ethnicity== c("ASN") & dates=="05")
index2 = which(ethnicity== c("CEU") & dates=="02")
index2<-sample(index2,3)
index<-c(index1,index2)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qvals<-qvalue(pvals)
print(sum(qvals$qvalues<.05))
```


## Exercises 3
```{r}
library(GSE5859Subset)
data(GSE5859Subset)
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
table(sampleInfo$group, month)
```
### 1. Using the functions rowttests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1?
```{r}
library(sva) #available from Bioconductor
mod <- model.matrix(~sex)
library(qvalue)
res <- rowttests(geneExpression,as.factor( sampleInfo$group ))
qvals <- qvalue(res$p.value)$qvalue
index <- which(qvals<0.1)
print(length(index))
```

### 2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX. We do not expect many autosomal genes to be different between males and females. This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY. For the list calculated above, what proportion of this list is on chrX or chrY?
```{r}
library(rafalib)
library(Biobase)
library(GSE5859) ##Available from GitHub
data(GSE5859)

cors <- cor(exprs(e))
Pairs=which(abs(cors)>0.9999,arr.ind=TRUE)
out = Pairs[which(Pairs[,1]<Pairs[,2]),,drop=FALSE]
if(length(out[,2])>0) e=e[,-out[2]]

out <- grep("AFFX",featureNames(e))
e <- e[-out,]

y <- exprs(e)-rowMeans(exprs(e))
dates <- pData(e)$date
eth <- pData(e)$ethnicity

annotation(e)
library(hgfocus.db) ##install from Bioconductor
annot <- select(hgfocus.db, keys=featureNames(e), keytype="PROBEID",
columns=c("CHR"))
##for genes with multiples, pick one
annot <-annot[match(featureNames(e),annot$PROBEID),]
annot$CHR <- ifelse(is.na(annot$CHR),NA,paste0("chr",annot$CHR))
##compute median expression on chromosome Y

newVec<-annot$CHR
head(annot$CHR)
sum(newVec[index]=="chrY",na.rm=TRUE)
sum(newVec[index]=="chrX",na.rm=TRUE)
```

Chromosome Y proportion is 8/59
Chromosome x proportion is 12/59
Therefore, total on Chromosome X or Y is 20

### 3.We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?

12


### 4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05 ?
```{r}


sex<-factor(sampleInfo$group)
pvals <- rowttests(geneExpression,sex)$p.value
qvals <- qvalue(pvals)$qvalue

index=which(qvals<.1 & !(annot$CHR%in%c("chrX","chrY")))
factor=factor(format(sampleInfo$date,"%m"))
res <- rowttests(geneExpression[index,],factor)
mean(res$p.value<.05)

```

.8717

### 5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix? . A) X = model.matrix(~sex+ethnicity) . B) X = cbind(sex,as.numeric(month)) . C) It can't be done with one line. . D) X = model.matrix(~sex+month)

Option D

### 6. Now use the X defined above, to fit a regression model using lm for each gene. You can obtain p-values for estimated parameters using summary. Here is an example
```{r}
X = model.matrix(~sex+month)
i = 234
y = geneExpression[i,]
fit = lm(y~X)
summary(fit)$coef
```
### How many of the q-values for the group comparison are now <0.1?
```{r}
X = model.matrix(~sex+month)
vectorPvalues<-vector(mode="double",length=0)
for (i in 1:nrow(geneExpression)){
  y = geneExpression[i,]
  fit = lm(y~X)
  vectorPvalues[[i]]<-summary(fit)$coef[2,4]
}
qvals <- qvalue(vectorPvalues)$qvalue
index=which(qvals<0.1)
newList<-vectorPvalues[index]
sum(qvals<0.1)
```
17 q-values are <.1

### 7. With this new list, what proportion of these are chrX and chrY? Notice the big improvement.
```{r}
index=which(qvals<0.1)
newVec<-annot$CHR
mySumXOrY<-sum(newVec[index]=="chrY",na.rm=TRUE)+sum(newVec[index]=="chrX",na.rm=TRUE)
newListLength<-17
proportion<-mySumXOrY/newListLength
print(proportion)
```


### 8. How many on Y or X?
```{r}
index=which(qvals<0.1)
mySumXOrY<-sum(newVec[index]=="chrY",na.rm=TRUE)+sum(newVec[index]=="chrX",na.rm=TRUE)
print(mySumXOrY)
```

### 9. Now from the linear model above, extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1?
```{r}
X = model.matrix(~sex+month)
vectorPvalues<-vector(mode="double",length=0)
for (i in 1:nrow(geneExpression)){
  y = geneExpression[i,]
  fit = lm(y~X)
  vectorPvalues[[i]]<-summary(fit)$coef[3,4]
}
qvals <- qvalue(vectorPvalues)$qvalue
index=which(qvals<0.1)
newList<-vectorPvalues[index]
print(sum(qvals<0.1))
```

## Excercises 4

### 1. Suppose you want to make an MA plot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?

Option B

### 2.Say Y is M X N, in the SVD Y = UDV ??? which of the following is not correct? . A) DV ??? are the new coordinates for the projection U???Y . B) UD are the new coordinates for the projection Y V . C) D are the coordinates of the projection U???Y . D) U???Y is a projection from an N-dimensional to M-dimensional subspace.

Option C

### 3. Define:
```{r}
library(Biobase)
library(GSE5859Subset)
data(GSE5859Subset)
y = geneExpression - rowMeans(geneExpression)
```
### Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted. Based on these plots, which of the following you would say is true?
```{r}

n <- nrow(pData(e))
o <- order(pData(e)$date)
Y1=exprs(e)
cors=cor(Y1-rowMeans(Y1))
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))

Y2=exprs(e)[,o]
cors=cor(Y2-rowMeans(Y2))
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```

Option C

### 4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates. Which command gives us these estimates? . A) pcs = svd(y)$v[1:2,] . B) pcs = svd(y)$v[,1:2] . C) pcs = svd(y)$u[,1:2] . D) pcs = svd(y)$d[1:2]
```{r}
y = geneExpression - rowMeans(geneExpression)
s <- svd(y)
W <- t(s$v[,1:2])
colnames(W)<-colnames(y)
round(W,2)
fit = s$u[,1:2]%*% (s$d[1:2]*W)
var(as.vector(fit))/var(as.vector(y))
```

Option B

### 5. Plot each of the estimated factors ordered by date. Use color to denote month. The first factor is clearly related to date. Which of the following appear to be most different according to this factor? . A) June 23 and June 27 . B) Oct 07 and Oct 28 . C) June 10 and June 23 . D) June 15 and June 24
```{r}
times <-sampleInfo$date
format(times,"%m")
mypar(1,1)
o=order(times)
times=as.numeric(times)[o]
pcs = svd(y)$v[,1:2]

plot(pcs[o,1],col=times,xaxt="n",xlab="")
label = gsub("2005-","",sampleInfo$date[o])
axis(1,1:ncol(y),label,las=2)

plot(pcs[o,2],col=times,xaxt="n",xlab="")
label = gsub("2005-","",sampleInfo$date[o])
axis(1,1:ncol(y),label,las=2)


```

Option A

### 6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?

```{r}
s <- svd(y)
mypar(1,1)
o<-order(times)
cols <- as.numeric(times)
plot(s$v[o,1],pch=21,cex=1.25,bg=cols[o],ylab="First PC",xaxt="n",xlab="")
legend("topleft",c("Month 1","Month 2"),col=1:2,pch=16,box.lwd=0)
mypar(1,1)
plot(s$d^2/sum(s$d^2),ylab="% variance explained",xlab="Principal component")
```

From the created plot, 2 PCs explain more than 10% of the variability.

### 7. Which PC most correlates (negative or positive correlation) with month?
```{r}
month = factor(format(sampleInfo$date,"%m"))
pcs = svd(y)$v
myVec<-abs(cor(pcs,as.numeric(month)))
which.max(myVec)
```

The first PC

### 8. What is this correlation (in absolute value)?

```{r}
month = factor(format(sampleInfo$date,"%m"))
pcs = svd(y)$v
myVec<-abs(cor(pcs,as.numeric(month)))
print(myVec[which.max(myVec)])
```

### 9. Which PC least correlates (negative or positive correlation) with sex?

```{r}
month = factor(format(sampleInfo$date,"%m"))
pcs = svd(y)$v
myVec<-abs(cor(pcs,as.numeric(month)))
which.min(myVec)
```

The 21st PC.


### 10. What is this correlation (in absolute value)?
```{r}
month = factor(format(sampleInfo$date,"%m"))
pcs = svd(y)$v
myVec<-abs(cor(pcs,as.numeric(month)))
print(myVec[which.min(myVec)])
```

### 11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?

```{r}
X = model.matrix(~sex+svd(y)$v[,1:2])
vectorPvalues<-vector(mode="double",length=0)
for (i in 1:nrow(geneExpression)){
  y = geneExpression[i,]
  fit = lm(y~X)
  vectorPvalues[[i]]<-summary(fit)$coef[2,4]
}
qvals <- qvalue(vectorPvalues)$qvalue
index=which(qvals<0.1)
newList<-vectorPvalues[index]
print(sum(qvals<0.1))

```

### 12. What proportion of the genes are on chromosomes X and Y?

```{r}
index=which(qvals<0.1)
newVec<-annot$CHR
mySumXOrY<-sum(newVec[index]=="chrY",na.rm=TRUE)+sum(newVec[index]=="chrX",na.rm=TRUE)
newListLength<-14
proportion<-mySumXOrY/newListLength
print(proportion)
```

1, or all of them, are on chromosomes X and Y


## Exercises 5

### 1. 
```{r}
svafit <- sva(geneExpression,mod)
svaX<-model.matrix(~sex+svafit$sv)
lmfit <- lmFit(geneExpression,svaX)
tt<- lmfit$coef[,2]*sqrt(lmfit$df.residual)/(2*lmfit$sigma)
res <- data.frame(dm= -lmfit$coef[,2],
p.value=2*(1-pt(abs(tt),lmfit$df.residual[1]) ) )
qvals <- qvalue(res$p.value)$qvalue
index <- which(qvals<0.1)
newVec<-annot$CHR
cat("Total genes with q-value < 0.1: ",length(index),"\n",
"Number of selected genes on chrY: ", sum(newVec[index]=="chrY",na.rm=TRUE),"\n\
",
"Number of selected genes on chrX: ", sum(newVec[index]=="chrX",na.rm=TRUE),sep="")
```
### 2. How many of these genes are from chrY or chrX?

Chromosome Y=5
Chromosome X=8